### PR TITLE
Add Manifold Base Mint Ingestor

### PR DESCRIPTION
**Pull request by polymawutor**
This Ingestor builds a `MintTemplate` from URL only. Manifold's Claims API allows us to query claims (mints) by Slug, and in return we get detailed metadata about mints. The only read we make onchain is to `getFeeBps` needed to calculate the total price to be paid. 

I would appreciate if you could take a look at the parameters for the mint function - `mintBaseNew` - which takes 3 parameters including an `uri`.

This PR works on Issue #9

### CODE REVIEW

**Comment by chrismaddern** [comment]
Thank you for this submission!! 🥳 🙏 

We've gotten a little behind on reviewing these, but will be spending some time over the next 48 hours, so look out for some feedback soon! Thank you! 🙏

**Comment by chrismaddern** [comment]
@mawutory I just went to review this one for you and it seems that this one erroneously includes an Opensea ingestor too?

Can you remove and then we can review this one and hopefully get it across the line quickly now that you know how to make one!!! 🙂

**Comment by mawutory** [comment]
yes right, i've been working on this one intermittently. let me clean it up and revert.

thanks!

**Comment by mawutory** [comment]
@chrismaddern i have fixed all issues, removed the opensea addition, and all tests and simulations passed!

looking forward to adding this one. thanks!

**Comment by chrismaddern** [comment]
Thanks @mawutory the tests are running!! 🥳 

I went to try this and the very first one failed:
https://app.manifold.xyz/c/seeking-self-finding-peace

This is an ETH manifold piece. We should _either_:

* Support ETH
* OR... limit supportsUrl / supportsContract to decline to ingest

It is an error for an ingestor to return `true` to supportsUrl, and then throw during ingestUrl

Supporting ETH is obviously great, but would be fine to just make the implementation of supportsUrl/contract to be correct

**Comment by mawutory** [comment]
@chrismaddern `supportsUrl` and `supportsContract` updated to decline incompatible URLs and contracts.

**Comment by chrismaddern** [comment]
@mawutory I'm not sure if I'm getting unlucky, if you just need to do a little more testing... the first base one I tried also doesn't work (I just searched twitter and picked the first base one)

I recommend try a half dozen or so before resubmitting so that we can get it approved quickly, vs. the back-and-forth.

https://app.manifold.xyz/c/bythebook

Really hoping to get this across the line.

File: src/ingestors/index.ts [hunk 1]
```diff
@@ -1,7 +1,16 @@
 import { FxHashIngestor } from './fxhash';
 import { MintIngestor } from '../lib/types/mint-ingestor';
 import { ProhibitionDailyIngestor } from './prohibition-daily';
+<<<<<<< HEAD
```
Comment (chrismaddern) [inline src/ingestors/index.ts:4]: It looks like this may have some git merge conflicts

**Comment by mawutory** [comment]
@chrismaddern 

support for ingesting via only contract address had to be removed because the only way to get the slug from contract address does not work for some type of mints, therefore only ingesting via URL can be supported.

File: src/ingestors/index.ts (hunk 1 above)
Comment (mawutory) [reply src/ingestors/index.ts:4]: yes, will take care of that.

**Comment by mawutory** [comment]
@chrismaddern please merge conflicts have been resolved.

**Comment by mawutory** [comment]
@chrismaddern please can you take a look at the latest change. here's a summary so far since the last review.

1. ingesting mints via URL only (no reliable way to build URL via contract address)
2. adopting contract parameters to dynamic instance IDs
3. getting owner address from contract (not available via API)
4. expired mints are denied one endDate is less than current date (previously expired mints used to proceed and failed at simulation)

live mints that you could test with:
- https://app.manifold.xyz/c/spaceexplorer
- https://app.manifold.xyz/c/solent

these mints will fail because they're either expired or not initialized
- https://app.manifold.xyz/c/black-oak-coffee
- https://app.manifold.xyz/c/bbnftcollectible
- https://app.manifold.xyz/c/satoshis-slices

non-initialized mints will fail at simulation with the message - "Claim not initialized". 

standard, working mints on manifold should pass.

**Comment by mawutory** [comment]
@chrismaddern Sir any update on this? I would appreciate some closure. Thanks!

**Comment by chrismaddern** [comment]
Thanks @polymawutor — I'm really excited to get this in and would love to do so in the next couple days and get it paid out if you're up for it!

A couple things:

* This did not properly support ERC721 on Base — it was hardcoded to ERC-1155. I've implemented this in the following patch file which you can apply:
[manifold_fixes.patch](https://github.com/user-attachments/files/16761359/manifold_fixes.patch)

* This does not support ingestion by contract currently

The Manifold creator contract supports a method `getExtensions` on the contract to get the extensions — we can call that, and check to see if it's one of our known mint addresses is in the extensions list - that tells us it's (1) Manifold and (2) we support it!

https://basescan.org/address/0xe2cf639a5eba5e8d1e291aeb44ac66c8c0727f98#readProxyContract

CleanShot 2024-08-27 at 07 23 51@2x

Let me know if you need any more help, but otherwise, seems this is very landable from here!

**Comment by chrismaddern** [comment]
(This was my test mint for 721 support btw - https://app.manifold.xyz/c/freedom-to-be-creative)

**Comment by polymawutor** [comment]
Hello @chrismaddern thanks for explaining and helping out with the patch! I appreciate it.

After applying the patch, I started doing more tests. One of the mints (https://app.manifold.xyz/c/basedpunks) is sold out but was able to pass and eventually failed at the simulation. 

As you mentioned earlier, we have to filter cases like this before ingesting.

So I added one more check to filter out sold-out mints. I used the `getClaim` method on both ERC1155 and ERC721 Claims contracts to get `total` and `totalMax` to infer whether a mint was sold out or not.

I also modified this error message in `index.ts` to reflect all the possible causes:
```
throw new MintIngestorError(MintIngestionErrorName.MissingRequiredData, 'Missing required data, mint expired, or sold out');
```

So far all my tests have passed. Here are all the mints I have tested so far:

**ERC1155**
- https://app.manifold.xyz/c/bythebook (live)
- https://app.manifold.xyz/c/spaceexplorer (live)
- https://app.manifold.xyz/c/solent (live)
- https://app.manifold.xyz/c/satoshis-slices (ended/expired)

**ERC721**
- https://app.manifold.xyz/c/seeking-self-finding-peace (eth mainnet, not supported)
- https://app.manifold.xyz/c/black-oak-coffee (ended/expired)
- https://app.manifold.xyz/c/bbnftcollectible (live)
- https://app.manifold.xyz/c/freedom-to-be-creative (live)
- https://app.manifold.xyz/c/basedpunks (ended/expired)

What do you think? Have we filtered out all edge cases?

**Comment by chrismaddern** [comment]
Amazing.

It does look like the automated tests are failing for the ingestion by contract?

CleanShot 2024-08-28 at 13 57 44@2x

It looks like you're passing a chainId/ contract pair to the contract test but then failing it because it doesn't have a URL?

**Comment by chrismaddern** [comment]
Really excited to get this in... I want to start using it immediately!

**Comment by polymawutor** [comment]
@chrismaddern Since ingestion by contract only is not supported, I should have removed that portion of the automated tests? Apologies, will push an update shortly.

**Comment by chrismaddern** [comment]
It definitely can support it though, right? 

> The Manifold creator contract supports a method getExtensions on the contract to get the extensions — we can call that, and check to see if it's one of our known mint addresses is in the extensions list - that tells us it's (1) Manifold and (2) we support it!
https://basescan.org/address/0xe2cf639a5eba5e8d1e291aeb44ac66c8c0727f98#readProxyContract

I left the image above. Would be really helpful to support ingestion by contract if it's possible -- most others do.

**Comment by polymawutor** [comment]
@chrismaddern Please how do we build the URL from there? Since we need the `Slug` or `instanceId` to query the API. The contracts don't provide neither.

**Comment by chrismaddern** [comment]
Okay, for today, let's make the tests pass, and then we can work on that after merge?

**Comment by polymawutor** [comment]
Sure.

**Review by chrismaddern — APPROVED** [review]
